distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Postgres/Pgpass.enso

```diff
@@ -0,0 +1,126 @@
+
+   If `PGPASSFILE` environment variable is set, the provided location is
+   used instead of the default one.
+
+   Arguments:
+   - host: The hostname of the database server.
+   - port: The port of the database server.
+   - database: The database to connect to.
+read : Text -> Integer -> Text -> Text -> [Pair Text Text]
+read host port database username=Nothing =
+    pgpass_file = locate
+    if pgpass_file.is_nothing || (verify pgpass_file . not) then [] else
+        entries = parse_file pgpass_file
+        found = entries.find entry->
+            entry.matches host port database username
+        case found.catch Nothing of
+            Nothing -> []
+            entry -> [Pair 'user' entry.username, Pair 'password' entry.password]
+
+type Pgpass_Entry
+    ## PRIVATE
+    type Pgpass_Entry host port database username password
+
+    ## PRIVATE
+    matches : Text -> Text|Integer -> Text -> Text -> Boolean
+    matches host port database username=Nothing =
+        wildcard='*'
+        host_match = self.host==wildcard || self.host==host
+        port_match = self.port==wildcard ||
+            normalized_port = case port of
+                Integer -> port.to_text
+                Text -> port
+            self.port==normalized_port
+        database_match = self.database==wildcard || self.database==database
+        username_match = username==Nothing || self.username==wildcard || self.username==username
+        host_match && port_match && database_match && username_match
+
+## PRIVATE
+   Determines the location of the .pgpass file to use.
+locate = case Environment.get "PGPASSFILE" of
+    Nothing -> case Platform.os of
+        Platform.Windows -> case Environment.get "APPDATA" of
+            Nothing -> Nothing
+            appdata -> File.new appdata / "postgresql" / "pgpass.conf"
+        _ -> case Environment.get "HOME" of
+            Nothing -> Nothing
+            home -> File.new home / ".pgpass"
+    path -> File.new path
+
+## PRIVATE
+   Checks if the given .pgpass file can be used.
+
+   The file can be used if it exists and has correct permissions on UNIX systems.
+verify file = case Platform.os of
+    Platform.Windows -> file.exists
+    _ -> case file.exists of
+        False -> False
+        True ->
+            permissions = file.posix_permissions
+            can_others_access = permissions.group.not_empty || permissions.others.not_empty
+            can_others_access.not
+
+## PRIVATE
+parse_file file =
+    parse line =
+        if line.starts_with "#" || line.is_empty then Nothing else
+            elements = parse_line line
+            if elements.length != 5 then Nothing else
+                Pgpass_Entry (elements.at 0) (elements.at 1) (elements.at 2) (elements.at 3) (elements.at 4)
+
+    File.read_text file . lines . map parse . filter (x -> x.is_nothing.not)
+
+## PRIVATE
+parse_line line =
+    existing_entries = Vector.new_builder
+    current_entry = Java_String_Builder.new
```

Comment:
Wow, `java.lang.StringBuilder`! `Text` has some support for _ropes_ (fast concatenation) - wouldn't that enough?

Reply:
Reason this is as mutating the builder - could be worked around by using a Ref and updating each go, but as internal StringBuilder feels easier.